pkg/kvstore/allocator/doublewrite/backend.go

```diff
@@ -0,0 +1,227 @@
+	crdKey, crdErr := d.crdBackend.AllocateID(ctx, id, key)
+	if crdErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(crdErr).Error("CRD backend failed to allocate identity")
+	}
+	kvStoreKey, kvStoreErr := d.kvstoreBackend.AllocateID(ctx, id, key)
+	if kvStoreErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(kvStoreErr).Error("KVStore backend failed to allocate identity")
+	}
+	if d.readFromKVStore {
+		return kvStoreKey, kvStoreErr
+	}
+	return crdKey, crdErr
+}
+
+func (d *doubleWriteBackend) AllocateIDIfLocked(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, lock kvstore.KVLocker) (allocator.AllocatorKey, error) {
+	crdKey, crdErr := d.crdBackend.AllocateIDIfLocked(ctx, id, key, lock)
+	if crdErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(crdErr).Error("CRD backend failed to allocate identity with lock")
+	}
+	kvStoreKey, kvStoreErr := d.kvstoreBackend.AllocateIDIfLocked(ctx, id, key, lock)
+	if kvStoreErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(kvStoreErr).Error("KVStore backend failed to allocate identity with lock")
+	}
+	if d.readFromKVStore {
+		return kvStoreKey, kvStoreErr
+	}
+	return crdKey, crdErr
+}
+
+func (d *doubleWriteBackend) AcquireReference(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, lock kvstore.KVLocker) error {
+	crdErr := d.crdBackend.AcquireReference(ctx, id, key, lock)
+	if crdErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(crdErr).Error("CRD backend failed to acquire reference with lock")
+	}
+	kvStoreErr := d.kvstoreBackend.AcquireReference(ctx, id, key, lock)
+	if kvStoreErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(kvStoreErr).Error("KVStore backend failed to acquire reference with lock")
+	}
+	if d.readFromKVStore {
+		return kvStoreErr
+	}
+	return crdErr
+}
+
+func (d *doubleWriteBackend) RunLocksGC(ctx context.Context, staleKeysPrevRound map[string]kvstore.Value) (map[string]kvstore.Value, error) {
+	// This is a no-op for the CRD backend
+	return d.kvstoreBackend.RunLocksGC(ctx, staleKeysPrevRound)
+}
+
+func (d *doubleWriteBackend) RunGC(
+	ctx context.Context,
+	rateLimit *rate.Limiter,
+	staleKeysPrevRound map[string]uint64,
+	minID, maxID idpool.ID,
+) (map[string]uint64, *allocator.GCStats, error) {
+	// This is a no-op for the CRD backend
+	return d.kvstoreBackend.RunGC(ctx, rateLimit, staleKeysPrevRound, minID, maxID)
+}
+
+func (d *doubleWriteBackend) UpdateKey(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, reliablyMissing bool) error {
```

Comment:
it seems like `reliablyMissing` isn't really used anymore?

Reply:
Yes, there is actually a thread about it here: https://github.com/cilium/cilium/pull/31920#discussion_r1650610739